src/main/java/frc/robot/subsystems/climb/ClimbIOSim.java

```diff
@@ -0,0 +1,79 @@
+package frc.robot.subsystems.climb;
+
+import static edu.wpi.first.units.Units.*;
+
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.system.plant.DCMotor;
+import edu.wpi.first.units.measure.Angle;
+import edu.wpi.first.units.measure.MutAngle;
+import edu.wpi.first.units.measure.MutVoltage;
+import edu.wpi.first.units.measure.Voltage;
+import edu.wpi.first.wpilibj.simulation.SingleJointedArmSim;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
+public class ClimbIOSim implements ClimbIO {
+
+    private SingleJointedArmSim climb =
+            new SingleJointedArmSim(
+                    DCMotor.getKrakenX60(2),
+                    75,
+                    SingleJointedArmSim.estimateMOI(0.5, 0.5),
+                    0.5,
+                    0,
+                    2 * Math.PI,
+                    false,
+                    0,
+                    0,
+                    0);
+
+    private final PIDController angleController = new PIDController(10, 0, 0);
+
+    private MutAngle goalAngle = Radians.mutable(0);
+    private MutVoltage overrideVoltage = Volts.mutable(0.0);
+
+    private boolean override = false;
+
+    public ClimbIOSim() {
+        SmartDashboard.putBoolean("lockedToCage", false);
+    }
+
+    @Override
+    public void updateInputs(ClimbInputs inputs) {
+
+        inputs.lockedToCage = SmartDashboard.getBoolean("lockedToCage", false);
+        inputs.goalAngle.mut_replace(goalAngle);
+        inputs.motorAngle.mut_replace(Radians.of(climb.getAngleRads()));
+    }
+
+    @Override
+    public void applyOutputs(ClimbOutputs outputs) {
+        Voltage appliedVolts;
+        if (override) {
+            appliedVolts = overrideVoltage;
+        } else {
+            appliedVolts =
+                    Volts.of(
+                            angleController.calculate(climb.getAngleRads(), goalAngle.in(Radians)));
+        }
+
+        outputs.appliedVoltage.mut_replace(appliedVolts);
+        climb.setInputVoltage(appliedVolts.in(Volt));
```

Comment:
I think you need to also call `climb.update(0.02)` where 0.02 is the update rate (we probably have a constant lying around somewhere for this).

I recommend taking a look at the example code for this here: https://github.com/wpilibsuite/allwpilib/blob/main/wpilibjExamples/src/main/java/edu/wpi/first/wpilibj/examples/armsimulation/subsystems/Arm.java

Reply:
Also, I think we should make a separate function for doing the simulation called updateSimulation() and then call it from either applyInputs or applyOutputs

The argument for doing it in applyInputs is that you can then use the sim to update sensor values